### ISSUE TITLE
Can we load the heatmap layer with color depend on the field of the data ?

### ISSUE THREAD

**Issue by Raymond1030** [mapbox/mapbox-gl-js#13029]
The document tell that we can set the color by its points' density but not adjust by the field.
I try to create a heatmap like this (color depend on the field of the data),
but it doen't work. 

It trigger the error: "Error: layers.gas-heat.paint.heatmap-color: data expressions not supported"
It seem the color of heatmap layer must be designed by the density.

`const gas = getJson('../data/h2s_data_geojson.json');
      gas.features.forEach(feature => {
        const h2sValue = feature.properties.concentration_mg_per_m3; 
        let color;
        if (h2sValue < 432) {
          color = 'green';
    
        } 
        else if(432<=h2sValue < 618) {
          color = 'yellow';
        }
        else {
          color = 'red';
        }
        // Add a new 'color' property to the feature
        feature.properties.color = color;
      });
      console.log(gas)
      map.addSource('gas', {
        'type': 'geojson',
        'data': gas
      });
    
      map.addLayer(
        {
          'id': 'gas-heat',
          'type': 'heatmap',
          'source': 'gas',
    
          'paint': {
            'heatmap-color': ['get', 'color'] 
            ,
            'heatmap-radius': {
              stops: [[11, 12], [15, 12]]
            },
            'heatmap-opacity': {
              default: 1,
              stops: [[14, 1], [17, 0]]
            },
          }
        },
      );`

**Comment by cscetbon** [comment]
@Raymond1030 did you make any progress or find a workaround ?

It would be great to have a solution even if it means to have to fork the current code until it's available. I see that only the expression `heatmap-density` is supported but there should be a way to accept to calculate the Inverse Distance Weighting of any expression passed, and so something like this would then be accepted 
```
heatmap-color': [
  'interpolate',
  ['linear'],
  ['get', 'mag'],
  0, 'rgba(0, 255, 0, 0)',
  0.1, 'rgba(0, 255, 0, 0.9)',
  0.5, 'rgba(0, 255, 0, 1)',
  0.7, 'rgba(255, 255, 0, 1)',
  0.9, 'rgba(255, 128, 0, 1)',
  0.95, 'rgba(255, 0, 0, 1)'
],
```

**Comment by cscetbon** [comment]
@rreusser Could you propose an easy way of doing so ? I have noticed that you worked on raster-color based on heatmap-color.

**Comment by Raymond1030** [comment]
> @Raymond1030 did you make any progress or find a workaround ?
> 
> It would be great to have a solution even if it means to have to fork the current code until it's available. I see that only the expression `heatmap-density` is supported but there should be a way to accept to calculate the Inverse Distance Weighting of any expression passed, and so something like this would then be accepted
> 
> ```
> heatmap-color': [
>   'interpolate',
>   ['linear'],
>   ['get', 'mag'],
>   0, 'rgba(0, 255, 0, 0)',
>   0.1, 'rgba(0, 255, 0, 0.9)',
>   0.5, 'rgba(0, 255, 0, 1)',
>   0.7, 'rgba(255, 255, 0, 1)',
>   0.9, 'rgba(255, 128, 0, 1)',
>   0.95, 'rgba(255, 0, 0, 1)'
> ],
> ```

@cscetbon No, I didn't fix it. I can't make it finally, so i just changed my mind to process data in ArcGIS. 
I'm trying your method, but it doesn't seem to be working.
Error:
`evented.js:151  Error: layers.gas-heat.paint.heatmap-color: data expressions not supported`

**Comment by cscetbon** [comment]
>I'm trying your method, but it doesn't seem to be working.
Error:
evented.js:151  Error: layers.gas-heat.paint.heatmap-color: data expressions not supported

@Raymond1030 yeah it's not supported today, I was just saying that if that feature was implemented (hypothetically) we could then use such a syntax.

> @cscetbon No, I didn't fix it. I can't make it finally, so i just changed my mind to process data in ArcGIS.

Do you have a gist that I can look at to see how it works ? But ArcGIS is not open source, right ?

**Comment by Raymond1030** [comment]
> >I'm trying your method, but it doesn't seem to be working.
> Error:
> evented.js:151  Error: layers.gas-heat.paint.heatmap-color: data expressions not supported
> 
> @Raymond1030 yeah it's not supported today, I was just saying that if that feature was implemented (hypothetically) we could then use such a syntax.
> 
> > @cscetbon No, I didn't fix it. I can't make it finally, so i just changed my mind to process data in ArcGIS.
> 
> Do you have a gist that I can look at to see how it works ? But ArcGIS is not open source, right ?

It didn't work in the effect what we hope. Because i just process them into line layer  (contour) for  visualization😂)

I'm not working on this right now, but I promise I can upload a gist and the data soon to test.

**Comment by Raymond1030** [comment]
@cscetbon Hi, I create a gist which provides the data and the scripts which I fail to render the data.
https://gist.github.com/Raymond1030/a6e95ea080ec2a1c94fd4b53a53ba490

Here is a small example of what I want to render like this.
<img width="372" alt="image" src="https://github.com/user-attachments/assets/e9e24e9d-84e5-4c3a-a36e-46596f1206b0">
This rendering depends on the field of the data (the concentration of the gas).

**Comment by cscetbon** [comment]
Trying again to poke @rreusser about it in case there could be a quick fix
